### PR TITLE
[MLIR] XeGPU dialect for Intel GPU - core definitions and base classes

### PR DESCRIPTION
**Pull request by chencha3**
This PR follows our previous [RFC ](https://discourse.llvm.org/t/rfc-add-xegpu-dialect-for-intel-gpus/75723) to add XeGPU dialect definition for Intel GPUs. It contains dialect, type, attributes and operators definitions, as well as testcases for semantic checks. The lowering and optimization passes will be issued with separated passes.

### CODE REVIEW

**Comment by Jianhui-Li** [comment]
@rengolin  @joker-eph  This is the XeGPU dialect definition. Could you please take a look and give us some guideline? Appreciate if you can invite some one to review it also.

**Comment by rengolin** [comment]
The code looks good to me. It's a GPU dialect like the others, it has load/store, MMA/SIMD ops, sub-group descriptors, etc. 

@adam-smnk

**Comment by rengolin** [comment]
@joker-eph what are the next steps here?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPU.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#include <mlir/Interfaces/SideEffectInterfaces.h>
+#include <mlir/Interfaces/ViewLikeInterface.h>
+
+namespace mlir {
+
+/// Return the list of Range (i.e. offset, size, stride). Each Range
+/// entry contains either the dynamic value or a ConstantIndexOp constructed
+/// with `b` at location `loc`.
+SmallVector<Range, 8> getOrCreateRanges(OffsetSizeAndStrideOpInterface op,
+                                        OpBuilder &b, Location loc);
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPU.h:31]: Seems like a general utility, not what I expect in the top-level dialect include file.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      The XeGPU dialect models Intel Xe ISA semantics but works at vector and
+      TensorDesc data type. It provides 1:1 mappings to match Xe instructions 
+      like DPAS and 2D block load. The matrix size being processed at this level
+      exactly matches the hardware instructions or the intrinsic supported by
+      the lower-level GPU compiler.
+    }];
+
+    let dependentDialects = [
+      "arith::ArithDialect",
+      "memref::MemRefDialect"
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Where are these dependencies coming from?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+    DefaultValuedAttr<XeGPU_ModeAttr, "xegpu::ModeKind::SIMT">: $mode);
+
+  let results = (outs XeGPU_TensorDesc: $result);
+
+  let hasCustomAssemblyFormat = 1;
+  let hasVerifier = 1;
+}
+
+def XeGPU_CreateDescOp: XeGPU_Op<"create_tdesc", [Pure]> {
+  let summary = "create scattered tensor descritors (TensorDesc).";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:227]:
```suggestion
  let summary = "create scattered tensor descriptors (TensorDesc).";
```

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+  let skipDefaultBuilders = 1;
+
+  // Format: xegpu.create_tdesc %src, %offsets {mode=simt, chunk_size_per_lane=1}
+  //              : ui64, vector<16 x index> -> !xegpu.tensor_desc<16xf32, #xegpu.scattered>
+  let hasCustomAssemblyFormat = 1;
+  let hasVerifier = 1;
+}
+
+def XeGPU_LoadGatherOp : XeGPU_Op<"load"> {
+  let summary = "load a scalar at source[offset].";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:275]: Is this op so trivial it does not need a documentation?
Seeing the  many arguments below I would think it is likely worth elaborating? (same for other ops in this file, please do another pass and reconsider them).

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+    VectorType getResultType() { 
+      return getResult().getType(); 
+    }
+  }];
+
+  let hasVerifier = 1;
+}
+
+def XeGPU_InvokeSIMDOp : XeGPU_Op<"invoke_SIMD", []> {
+    let summary = "Invoke_SIMD operation";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:417]: The summary isn't helpful here.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    let description = [{
+      The `xegpu.invoke_SIMD` operation works similar to a direct call to a function. 
+      But it is special to Intel GPU.
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:420]: The description "it is like call, but special" would deserve more details.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+          I8: $nbarrier_role,
+          I8Attr: $num_producers,
+          I8Attr: $num_consumers,
+          DefaultValuedAttr<XeGPU_ModeAttr, "xegpu::ModeKind::SIMT">: $mode);
+  let results = (outs XeGPU_Nbarrier: $result);
+  let hasCustomAssemblyFormat = 1;
+}
+
+def XeGPU_NbarrierArriveOp: XeGPU_Op<"nbarrier_arrive", []> {
+  let summary = "arrive at a named barrier.";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:481]: Does it mean "wait"?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+    OpBuilder<(ins "Type": $result, "xegpu::AtomicRMWKind": $kind,
+                   "Value": $tensorDesc, "Value": $mask, "Value": $value)>
+  ];
+
+  let skipDefaultBuilders = 1;
+  let hasVerifier = 1;
+}
+
+def XeGPU_AllocNbarrierOp: XeGPU_Op<"alloc_nbarrier", []> {
+  let summary = "allocate a specific number of named barriers.";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:463]: Where is the concept of a barrier or a named barrier documented?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+  let arguments = (ins I64Attr: $nbarrierCount);
+  let assemblyFormat = "$nbarrierCount attr-dict";
+}
+
+
+def XeGPU_CreateNbarrierOp: XeGPU_Op<"create_nbarrier", []> {
+  let summary = "create a named barrier.";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:470]: What is the difference between "Create" and "alloc" (above).
This goes with the overall question of documenting concepts like the barrier.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+  let assemblyFormat = [{ $payload attr-dict `:` qualified(type($payload)) }];
+}
+
+def XeGPU_CompileHintOp: XeGPU_Op<"compile_hint", []> {
+  let summary = "prevents the compiler from scheduling.";
+  let assemblyFormat = [{ attr-dict }];
+}
+
+def XeGPU_MfenceOp: XeGPU_Op<"mfence", []> {
+  let summary = "lsc fence.";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:498]: The summary is cryptic, and this is one more instance where it deserve some doc.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+def XeGPU_NbarrierWaitOp: XeGPU_Op<"nbarrier_wait", []> {
+  let summary = "wait for a named barrier.";
+  let arguments = (ins XeGPU_Nbarrier: $payload);
+  let assemblyFormat = [{ $payload attr-dict `:` qualified(type($payload)) }];
+}
+
+def XeGPU_CompileHintOp: XeGPU_Op<"compile_hint", []> {
+  let summary = "prevents the compiler from scheduling.";
+  let assemblyFormat = [{ attr-dict }];
+}
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:495]: I have concerns with the (undefined) semantics of this op right now, can you remove it and introduce it separately at a later time?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    * memory_scope (xegpu::MemoryScope): [optional] where the data is located, global memory or shared memory. It is default to Global.
+    * array_length (int): [optional] The number of continuous blocks with size as `shape`,
+               that will be loaded by block load at a time. It is default to 1.
+    * boundary_check (bool): [optional] indicates whether the operation detects the boundary and pads with zero for out-of-boundary access (default)
+    * scattered (xegpu::ScatteredAttr): [optional] It is a unit attribute. It can be only set as empty or ScatteredAttr, indicating
+                whether the TensorDesc is blocked (empty, default) or scattered (ScatteredAttr). If it is
+                blocked, rows are continuous in the correspoding dimention, otherwise, rows may be not continous.
+    * mapping (xegpu::SubGroupMapAttr): [optional] Used to guide compiler to distribute the workload into different threads. It is default to none.
+
+    For convinience, its attribute field can also take either "ScatteredAttr" or "SubGroupMapAttr" directly if and only
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td:64]:
```suggestion
    For convenience, its attribute field can also take either "ScatteredAttr" or "SubGroupMapAttr" directly if and only
```

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    xegpu::ScatteredAttr getScattered();
+    xegpu::SubGroupMapAttr getMapping();
+  }];
+
+  let hasCustomAssemblyFormat = true;
+}
+
+
+def XeGPU_Nbarrier: XeGPUTypeDef<"Nbarrier", "nbarrier", [], "mlir::Type"> {
+  let summary = "!xegpu.nbarrier a custom XeGPU type representing a barrier.";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUTypes.td:161]: Seems like maybe where everything about barriers should be documented, a bit like the tensordesc above.

File: mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+#include <mlir/Dialect/XeGPU/IR/XeGPU.cpp.inc>
+      >();
+  addAttributes<
+#define GET_ATTRDEF_LIST
+#include <mlir/Dialect/XeGPU/IR/XeGPUAttrs.cpp.inc>
+      >();
+}
+
+bool printDefaultValues() {
+  auto *env = getenv("MLIR_XEGPU_PRINT_DEFAULTS");
```
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:43]: MLIR dialect's behavior isn't sensitive to the environment right now, please remove and replace with regular setter/getter that a client can setup how they wish.

File: mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+      scattered = mlir::FieldParser<xegpu::ScatteredAttr>::parse(parser);
+      if (mlir::failed(scattered))
+        return parser.emitError(
+            parser.getCurrentLocation(),
+            "Failed to parse 'scattered' attr of TensorDescAttr, which is to "
+            "be a `xegpu::ScatteredAttr`");
+      seen_scattered = true;
+    }
+    return mlir::success();
+  };
```
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:150]: This looks like ODS generated code that has been slightly tweaked?

**Review by joker-eph — COMMENTED** [review]
Thanks for the patch. I don't have significant concerns with anything right now.

I would wonder if you could split this though and introduce this in a few steps?

I could imagine for example:

1) The basic dialect structure
2) Basic operations
3) The TensorDesc type and the operations interacting with it.
4) The barrier type and the operations interacting with it.


But you may see other independent blocks?

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td (hunk 2 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: I initially thought dependency exists when the dialect can be mixed with others. It looks like I was wrong. What does dependency among dialects mean? I failed to find a good example to help me understand it from the doc and in tree code. I saw SCF dialect depends on arith dialect. But how?

Comment (joker-eph) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Dependencies are used when you must load the other dialect to use this one.
For example that can come from folding, constant materialization, or canonicalization patterns.

Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: Thanks very much for your explanation. I removed them now.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td (hunk 4 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:275]: I updated its doc.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td (hunk 5 above)
Comment (chencha3) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUOps.td:417]: This op is temporary removed.

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUAttrs.td [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+  let builders = [
+    AttrBuilder<(ins
+      "llvm::ArrayRef<int32_t>":$wiLayout,
+      "llvm::ArrayRef<int32_t>":$wiData
+    )>
+  ];
+}
+
+def XeGPU_TensorDescAttr: XeGPUAttr<"TensorDesc", "tdesc_attr"> {
+  let parameters = (ins
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/XeGPU/IR/XeGPUAttrs.td:50]: We need some more comprehensive documentation for all types/attr (ideally with example syntax, etc.)

File: mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td (hunk 2 above)
Comment (joker-eph) [reply mlir/include/mlir/Dialect/XeGPU/IR/XeGPUDialect.td:39]: still there right now?

**Comment by chencha3** [comment]
> Thanks for the patch. I don't have significant concerns with anything right now.
> 
> I would wonder if you could split this though and introduce this in a few steps?
> 
> I could imagine for example:
> 
> 1. The basic dialect structure
> 2. Basic operations
> 3. The TensorDesc type and the operations interacting with it.
> 4. The barrier type and the operations interacting with it.
> 
> But you may see other independent blocks?

Thanks for your feedback. We plan to split it into following 5 steps according to your suggestions:

1. A PR simply contains definitions of dialect and base classes for op, type and attribute. We have updated this PR for this purpose.
2. A PR for TensorDesc type and attribute definitions and related operations including `CreateNdDescOp`, `LoadNdOp`, `StoreNdOp`, `PrefetchNdOp`, and `UpdateNdOffsetOp`. 
3. A PR for operators working on scattered data elements including `CreateDescOp`, `LoadGatherOp`, `StoreScatterOp`, `PrefetchOp`, `UpdateOffsetOp`.
4. A PR contains Named Barrier definition and related operators, including `nbarrier_alloc`, `create_nbarrier`, `nbarrier_arrive` and `nbarrier_wait`. 
5. A PR contains the rest of operators: including `DpasOp`, `MfenceOp`, `AtomicRMWOp` and `CompileHintOp`. 

Does this plan sound good to you?

**Comment by joker-eph** [comment]
Looks fine, thanks!

**Comment by chencha3** [comment]
> Looks fine, thanks!

Thanks Mehdi, I have trimmed this PR to only have core definitions and base classes. Does it look good to you? I will submit the PR for the second part when this one is approved.

**Comment by chencha3** [comment]
@joker-eph Hi Mehdi, what is my next step? merge this one first after getting approvals or submit other PRs first so they can be reviewed together?

**Comment by joker-eph** [comment]
You can pipeline and stack multiple PRs. However this requires you to push your branches to the LLVM repo, not open PRs from your fork (otherwise you can't cascade them).
It's too late for this one, but for the future ones :)

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+#include <mlir/Dialect/Linalg/IR/Linalg.h>
+#include <mlir/Dialect/MemRef/IR/MemRef.h>
+#include <mlir/Dialect/Tensor/IR/Tensor.h>
+#include <mlir/Dialect/Utils/StaticValueUtils.h>
+#include <mlir/Dialect/XeGPU/IR/XeGPU.h>
+#include <mlir/IR/Builders.h>
+#include <mlir/IR/DialectImplementation.h>
+#include <mlir/IR/TypeUtilities.h>
+#include <numeric>
+#include <type_traits>
```
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:21]: Can you clean up the include and add the required one with each PRs?

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+#define DEBUG_TYPE "xegpu"
+
+namespace mlir {
+class Token;
```
Comment (joker-eph) [inline mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:26]: Same for the fwd decl

**Review by joker-eph — APPROVED** [review]

File: mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp (hunk 18 above)
Comment (chencha3) [reply mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:26]: sure, thanks very much for your feedback.

**Review by rengolin — APPROVED** [review]
LGTM too, with @joker-eph comments addressed. Thanks!

**Comment by chencha3** [comment]
> LGTM too, with @joker-eph comments addressed. Thanks!
Thanks @rengolin, the code has now been cleaned.

**Comment by chencha3** [comment]
> You can pipeline and stack multiple PRs. However this requires you to push your branches to the LLVM repo, not open PRs from your fork (otherwise you can't cascade them). It's too late for this one, but for the future ones :)

Hi @joker-eph, not sure whether this is the right way to contact you though, here is PR for the second part (https://github.com/llvm/llvm-project/pull/84692). I don't have permission to push branch into the llvm-project repo, so I submitted PR from my fork for this one again. How can I apply for the access permission for future access?

**Comment by joker-eph** [comment]
Here are the info: https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access

**Pull request merged**